### PR TITLE
Remove ALT_JIT define and replace with runtime logic

### PR DESCRIPTION
**Pull request by davidwrighton**
The ALT_JIT define was causing confusion and problems when used in crossgen2
- It interfered with capturing jit dumps
- Didn't translate nicely to CoreRT codebase

This change replaces all of the existing #ifdef logic with a dynamic check of a jit flag.

### CODE REVIEW

**Comment by davidwrighton** [comment]
@dotnet/jit-contrib 
@BruceForstall, I know we discussed getting rid of the ALT_JIT ifdef. Is this the approach you were thinking of? Also, as I don't actually know what all the scenarios are for altjit, I need someone on the JIT team to test them for me.

**Comment by echesakovMSFT** [comment]
@dotnet/jit-contrib Assume I understand the main value of altjits - reproducing the assertions on cross platforms (x64 -> arm64, x86 -> arm, win-x64 -> linux-x64) I wonder if it is worth defining the altjit related checks under `DEBUG` ifdefs in order to minimize impact on the release jit.

> @davidwrighton Also, as I don't actually know what all the scenarios are for altjit, I need someone on the JIT team to test them for me.

We stopped running altjit in the CI a while ago. You can run the tests locally to verify the changes. 
You would need to specify:
```
COMPlus_AltJitName=<altjit.dll>
COMPlus_AltJit=*
```

**Comment by BruceForstall** [comment]
This should fix https://github.com/dotnet/runtime/issues/41643

We've used altjits in the past for:
1. reproducing asserts or debugging the JIT for cross-platform scenarios where it's easier to debug on some platforms (Windows x64) than others (arm64/32). This is the most common case.
2. bringing up a new compiler, either same arch or cross arch. E.g., bringing up RyuJIT was done this way as an altjit over JIT32/64.
3. for SPMI to avoid moving this existing JIT dll. IMO, this is a hack that has caused lots of problems and confusion.
4. to ship the RyuJIT CTPs.
5. for debugging same-platform JIT and using the COMPlus_AltJit options to enable your change for a subset of code. Could be useful for binary searching between your changes and a baseline.

**Comment by CarolEidt** [comment]
My main use of altjits is in running diffs, since it's much easier to use a single platform to get diffs for multiple targets.

**Comment by davidwrighton** [comment]
I think my changes address everything but the SPMI scenario. Bruce, could you elaborate on that one and give me some guidance on how one might handle that case?

**Comment by BruceForstall** [comment]
For SPMI, we need to introduce the SPMI shim between the JIT and VM. Traditionally, we would simply move the clrjit.dll someplace else, then copy the shim to the VM location as clrjit.dll, setting some variables so the shim can find the original clrjit.dll. It's annoying to move / rename files, so now we set the altjit variables to the SPMI shim, and then the shim loads the clrjit.dll (or potentially some cross-plat compiler) in its usual place. The annoying thing is we capture the COMPlus variables into the replay log, and we need to force them off when doing a SPMI replay.

File: src/coreclr/src/vm/jitinterface.cpp [hunk 1]
```diff
@@ -12260,6 +12258,8 @@ CorJitResult invokeCompileMethodHelper(EEJitManager *jitMgr,
 #endif
        )
     {
+        CORJIT_FLAGS altJitFlags = jitFlags;
+        altJitFlags.Set(CORJIT_FLAGS::CORJIT_FLAG_ALT_JIT);
```
Comment (BruceForstall) [inline src/coreclr/src/vm/jitinterface.cpp:12262]: `altJitFlags` is never used. The JIT is going to call back to get the flags. You need:
```
comp->setJitFlags(altJitFlags);
```

**Comment by AndyAyersMS** [comment]
`jit-diff-pmi` does a similar swapping of jits (moving checked base/diff jits in place of the base release jit), and allowing a normal jit to function as an altjit would simplify the workflow there too.

File: src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilationBuilder.cs [hunk 2]
```diff
@@ -74,9 +74,6 @@ public override CompilationBuilder UseBackendOptions(IEnumerable<string> options
                 builder.Add(new KeyValuePair<string, string>(name, value));
             }
 
-            // As we always use an AltJit to compile, tell the jit to always generate code
-            builder.Add(new KeyValuePair<string, string>("AltJitNgen", "*"));
```
Comment (BruceForstall) [inline src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilationBuilder.cs:78]: Yay!

**Comment by davidwrighton** [comment]
@AndyAyersMS This change makes all jits capable of acting like an altjit, on a per-method basis. From what you say, this should just work then.

@BruceForstall So, for SPMI, the fix would be to ignore the altjit jit flag when it comes from the VM, and add some other sort of config flag for SPMI to identify if it should treat the jit it wraps as being an alt jit? Is that what we would need here?

File: src/coreclr/src/jit/error.h [hunk 3]
```diff
@@ -153,24 +154,17 @@ extern void RecordNowayAssertGlobal(const char* filename, unsigned line, const c
...
 extern void notYetImplemented(const char* msg, const char* file, unsigned line);
 #define NYIRAW(msg) notYetImplemented(msg, __FILE__, __LINE__)
 
-#else // !defined(ALT_JIT)
-
-#define NYIRAW(msg) NOWAY_MSG(msg)
-
-#endif // !defined(ALT_JIT)
-
 #define NYI(msg)                    NYIRAW("NYI: " msg)
```
Comment (BruceForstall) [inline src/coreclr/src/jit/error.h:168]: We really should define NYI as a compile-time error for "complete" platforms (which now means all of them), but that's a separate issue.

**Review by BruceForstall — CHANGES_REQUESTED** [review]
Generally looks great. Thanks for doing this!

File: src/coreclr/src/jit/error.h [hunk 4]
```diff
@@ -153,24 +154,17 @@ extern void RecordNowayAssertGlobal(const char* filename, unsigned line, const c
 #define unreached() noWayAssertBody()
 
 #define NOWAY_MSG(msg) noWayAssertBodyConditional(NOWAY_ASSERT_BODY_ARGUMENTS)
+#define NOWAY_MSG_FILE_AND_LINE(msg, file, line) noWayAssertBodyConditional(NOWAY_ASSERT_BODY_ARGUMENTS)
```
Comment (BruceForstall) [inline src/coreclr/src/jit/error.h:157]: I guess `file` and `line` are ignored here, in favor of `__FILE__`, `__LINE__`

**Comment by BruceForstall** [comment]
> So, for SPMI, the fix would be to ignore the altjit jit flag when it comes from the VM, and add some other sort of config flag for SPMI to identify if it should treat the jit it wraps as being an alt jit? Is that what we would need here?

That would work. I'm not sure we need to worry about the scenario where SPMI collects using an altjit, but if so, maybe we need SPMI-specific flags to set altjit in that case.

I think we need a way to tell the VM to load a different primary JIT -- I don't want the SPMI shim to be loaded as an "altjit". We actually have `COMPlus_JitName` which we should use in `ExecutionManager::GetJitName()`. I fact, I think that's exactly what it used to do in desktop, but it appears in core someone removed the usage even though `COMPlus_JitName` still lives in clrconfigvalues.h.

File: src/coreclr/src/jit/error.h (hunk 4 above)
Comment (davidwrighton) [reply src/coreclr/src/jit/error.h:157]: Yes. Well, sortof. This is the release portion of the ifdef, where __FILE__ and __LINE__ aren't used at all. The goal is to match the overall behavior of what existed before.

File: src/coreclr/src/inc/corjitflags.h [hunk 5]
```diff
@@ -97,7 +97,7 @@ class CORJIT_FLAGS
 
         CORJIT_FLAG_NO_INLINING             = 42, // JIT should not inline any called method into this method
 
-        CORJIT_FLAG_UNUSED14                = 43,
+        CORJIT_FLAG_ALT_JIT                 = 43, // JIT should consider itself an ALT_JIT
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/coreclr/src/inc/corjitflags.h:100]: Ok

File: src/coreclr/src/jit/error.h (hunk 4 above)
Comment (davidwrighton) [reply src/coreclr/src/jit/error.h:157]: In particular, non-altjits don't put in file/line number stuff, but chk/dbg will.

**Comment by BruceForstall** [comment]
I built and tested basic JIT and NGEN altjit asm dumps (with the flags fix I showed) and it works fine.

Can you add the following to `CodeGen::genGenerateMachineCode`:
```
        if (compiler->opts.jitFlags->IsSet(JitFlags::JIT_FLAG_ALT_JIT))
        {
            printf("; invoked as altjit\n");
        }
```
at the end of the `if (compiler->opts.disAsm)` case? Then we can see in the asm output that this was an altjit compile.

Note also that you don't need the `opts.compAltJitRequested` member variable at all: you can always instead use the expression `opts.jitFlags->IsSet(JitFlags::JIT_FLAG_ALT_JIT)`.

**Comment by BruceForstall** [comment]
It looks to me like we need a change to SuperPMI to match this change.

Currently, when we do SPMI collection, we do so by setting AltJit to the SPMI shim. This causes the collection to capture the COMPlus_AltJit variables. On replay, we generally pass the (annoying) options `-jitoption force AltJit= -jitoption force AltJitNgen= ` to force the environment to set these variables to empty (overriding the collection), despite the fact they are defined in the collected environment variables. However, that doesn't work anymore, because the altjit flag is captured and replayed in the JIT flags, and our `-jitoption` to override the environment does not override the jit flags. So, we will always tell the replay JIT that it's an altjit. However, since `COMPlus_AltJit` is forced empty, the "altjit" will always say it isn't responsible for compiling the replay.

This implies something about how this is implemented. Currently, the JIT looks to see if `COMPlus_AltJit` is set, and if so, if the function we are compiling now matches what is specified, and sets `opts.altJit = true` if so. If `COMPlus_AltJit=*`, it matches any function (this is the only option for release builds). If `COMPlus_AltJit=MyMain`, it only invokes the altjit on MyMain, otherwise fails with CORJIT_SKIPPED. But presumably, all this logic should happen in the VM since if we get invoked with the new altjit jit flag, we should assume we're going to compile it in altjit "mode". And SPMI should probably just clear the altjit flag bit before invoking the JIT.

**Comment by BruceForstall** [comment]
@davidwrighton I submitted a PR that fixes the SPMI issue: https://github.com/davidwrighton/runtime/pull/1

@sandreenko Take a look; my PR is kind of a hack. I think we need to rethink how SPMI collections get invoked w.r.t. the altjit flags.

**Comment by sandreenko** [comment]
> @sandreenko Take a look; my PR is kind of a hack. I think we need to rethink how SPMI collections get invoked w.r.t. the altjit flags.

Do we set altjit flag during collection for CoreRun nowadays? For crossgen2 we pass a path to the invoked jit, so there we don't use altjit flags.

**Comment by BruceForstall** [comment]
> Do we set altjit flag during collection for CoreRun nowadays?

Yes, in both superpmi.py and the src\tests\JIT\superpmi>src\tests\JIT\superpmi\superpmicollect.cs, we set it for collection, and force clear it on replay.

**Comment by davidwrighton** [comment]
@BruceForstall I've merged your change, and responded to feedback. Please re-review.

**Review by BruceForstall — APPROVED** [review]
LGTM

**Comment by BruceForstall** [comment]
Don't you need to rebase instead of merge (from master)?

**Review by jkotas — APPROVED** [review]
Nice!

**Pull request merged**